### PR TITLE
add possibility to preload configured number of pages

### PR DESCRIPTION
**Pull request by pyatyispyatil**
Hello.

I'm not sure my approach to the work with coockies is correct. And I will be glad to get some recomendations about how to make it better.

### CODE REVIEW

File: src/content.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ function groupByRepo(events) {
 	return Array.from(events).reduce((repos, originalEvent) => {
 		const icon = originalEvent.querySelector('svg');
 		const actorEl = originalEvent.querySelector('.title a:first-child');
-		const eventEl = originalEvent.querySelector('.title a:nth-child(2)');
+		const eventEl = originalEvent.querySelector('.title a:last-child');
```
Comment (fregante) [inline src/content.js]: Why was this changed?

File: extension/manifest.json [hunk 2]
```diff
@@ -6,7 +6,9 @@
   "author": "Federico Brigante",
   "default_locale": "en",
   "permissions": [
-    "storage"
+    "storage",
+    "cookies",
```
Comment (fregante) [inline extension/manifest.json]: I think you don't have to manually specify the cookies here and in `fetch`, they should already be sent with `credentials: 'include'`: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

File: extension/manifest.json [hunk 3]
```diff
@@ -6,7 +6,9 @@
...
+    "https://github.com/"
```
Comment (fregante) [inline extension/manifest.json]: The permission `"https://github.com/"` is implied by the `content_scripts`, it shouldn't be necessary.

File: extension/options/index.html [hunk 4]
```diff
@@ -105,6 +105,23 @@ <h3 data-i18n="display_options_title"></h3>
...
+			<option value="5">5</option>
+			<option value="6">6</option>
+			<option value="7">7</option>
+			<option value="8">8</option>
+			<option value="9">9</option>
+			<option value="10">10</option>
+			<option value="11">11</option>
+		</select>
+	</label>
+
```
Comment (fregante) [inline extension/options/index.html]: This can just be a `<input type=number min=0>`, not a select

File: src/content.js [hunk 5]
```diff
@@ -53,8 +53,8 @@ function groupByRepo(events) {
 		return repos;
 	}, new Map());
 }
-function apply(options, insertionPoint) {
-	const holder = domifyEscape`<div class="ghcf-holder"><i>`;
+function apply(options) {
+	const holder = document.createDocumentFragment();
```
Comment (fregante) [inline src/content.js]: This change is unrelated and unwanted, although I don't remember why I had that `<i>` there.

File: src/content.js [hunk 6]
```diff
@@ -174,11 +188,78 @@ function apply(options, insertionPoint) {
...
+	})
+		.then(data => {
+			const streem = data.body;
+
+			return readAllChunks(streem)
+				.then(data => data
+					.map(chunk => new TextDecoder('utf-8').decode(chunk))
+					.join('')
+				);
+		});
```
Comment (fregante) [inline src/content.js]: You don't need all of this manual stream reading, there's a method to get the text:

```js
fetch(...).then(data => data.text())
```

File: src/content.js [hunk 7]
```diff
@@ -174,11 +188,78 @@ function apply(options, insertionPoint) {
...
+				);
+
+				preloadedNews.appendChild(fragment);
+				preloadedNews.appendChild(apply(options));
+
+				form.action = form.action.replace(/\/\d$/, '/' + (options.preloadPagesCount));
+			});
+	}
+}
+
```
Comment (fregante) [inline src/content.js]: I'd prefer letting GitHub load the pages so we can limit all of this to just `button.click()` similar to how refined-github does as you scroll down the page: https://github.com/sindresorhus/refined-github/blob/be756e68dd4bf24d53e802e1fec126506979fa80/extension/content.js#L150-L164

I know this means that the pages will be loaded serially instead of in parallel but:
- this way you don't have to wait for **all** the pages to be loaded before proceeding
- you can only ever see 1-2 pages per viewport, so loading more than that at once feels like too much.

**Review by fregante — CHANGES_REQUESTED** [review]
I like the idea but I think it can be done more easily. If you can make the requested changes perhaps it's best to start from scratch.

Perhaps refined-github's _download as you scroll_ behavior is enough for you as well.

File: src/content.js (hunk 1 above)
Comment (pyatyispyatil) [reply src/content.js]: This caused a bug when occurs element with one child. I tried to understand how this case could have been born, but could not. =(

File: src/content.js (hunk 5 above)
Comment (pyatyispyatil) [reply src/content.js]: I`ts still there in
```
function wrapHolder()
````

File: src/content.js (hunk 7 above)
Comment (pyatyispyatil) [reply src/content.js]: My feed with active filters looks like this:
oaum14fimgi 1
I think that this happens to many users and this feature can make life more easier. And small timeout in start is not so terrible.

And I think maybe it's worth to rewrite all the mechanics of the github news page and make them in a completely different form with sorting, global groups, tracking only certain events of a particular person or repository.  And generally abandon the standard GitHub's render.

File: extension/manifest.json (hunk 2 above)
Comment (pyatyispyatil) [reply extension/manifest.json]: Yes. I thought that working with cookies in an active session is not possible because of security policies. But it turned out that I just forgot about the fetch option. =)

File: extension/options/index.html (hunk 4 above)
Comment (pyatyispyatil) [reply extension/options/index.html]: GitHub allows you to request only 10 pages. And to use a input for the selected number in the range 1 - 10 is not usable. I already tried to do so.

File: src/content.js (hunk 5 above)
Comment (fregante) [reply src/content.js]: Yeah but it's just adding [extra code](https://github.com/gloooom/github-clean-feed/blob/0390edb8631ae078164cdc2eb3eb53ed30ee8056/src/content.js#L169-L180) to do the same thing, please restore

File: extension/options/index.html (hunk 4 above)
Comment (fregante) [reply extension/options/index.html]: Ok. I fixed the options UI in 25368e7.
Fetch my commit on your branch and then run `npm install` again.

File: src/content.js (hunk 7 above)
Comment (fregante) [reply src/content.js]: The empty feed in the screenshot is a separate issue, I create it here so maybe I'll fix it separately: https://github.com/bfred-it/github-clean-feed/issues/7

> generally abandon the standard GitHub's render.

That's a good point, worth discussing here: https://github.com/bfred-it/github-clean-feed/issues/8

**Review by fregante — APPROVED** [review]

File: src/content.js (hunk 5 above)
Comment (fregante) [reply src/content.js]: I see what the reason for this is: avoiding `Page X`. I want to keep that for now and solve #7 instead.

Also I remember that I used `<i>` to trick GitHub's own CSS to add a line above the first element:

screen shot 2017-06-07 at 19 45 28

File: src/content.js [hunk 8]
```diff
@@ -174,11 +188,44 @@ function apply(options, insertionPoint) {
...
+	})
+		.then(data => data.text());
+}
+
+function preloadPages(options) {
+	Promise.all(
+		Array(options.preloadPagesCount - 1)
+			.fill(null)
+			.map((t, n) => requestPage(n + 1))
+	)
```
Comment (fregante) [inline src/content.js]: Ok we can leave this custom fetching method, but I'd still like them to display as soon as possible, instead of waiting for all of them to load.

I'm not sure of the exact method, but:

* load them all at once like you're doing
* append the them as they come **but** maintain the correct order.

One of these functions will likely do that: https://github.com/sindresorhus/promise-fun

File: src/content.js [hunk 9]
```diff
@@ -174,11 +188,44 @@ function apply(options, insertionPoint) {
...
+		.then(data => {
+			const form = document.querySelector('.ajax-pagination-form');
+			const preloadedNews = document.querySelector('.ghcf-holder');
+
+			const fragment = document.createRange().createContextualFragment(
+				data.join('').replace(/<form[\s\S]*?<\/form>/g, '')
+			);
```
Comment (fregante) [inline src/content.js]: Try using the existing `domify` instead

**Review by fregante — CHANGES_REQUESTED** [review]

File: src/content.js (hunk 5 above)
Comment (pyatyispyatil) [reply src/content.js]: > Also I remember that I used <i> to trick GitHub's own CSS to add a line above the first element:

Another reason to switch to the [template](https://github.com/bfred-it/github-clean-feed/issues/8)

Well, I'll try to restore it.

File: src/content.js (hunk 8 above)
Comment (pyatyispyatil) [reply src/content.js]: Okay, I'll try to do.

File: extension/options/index.html (hunk 4 above)
Comment (pyatyispyatil) [reply extension/options/index.html]: Oh, it seems you broke the options. I see this after the update.
ed2d81a30b
And the extension console has no errors.

**Comment by bfred-it** [comment]
I'm working on this PR now, leave as is :)

**Comment by fregante** [comment]
This is more like what I had in mind, much less code necessary. ⬆️
